src/sd/spi.rs

```diff
@@ -0,0 +1,102 @@
+use core::borrow::Borrow;
+use core::marker::PhantomData;
+use core::ops::Deref;
+use core::sync::atomic::{AtomicU8, Ordering};
+
+use crate::gpio::{InputPin, OutputPin};
+use crate::peripheral::Peripheral;
+use crate::spi::SpiDriver;
+use crate::sys::*;
+
+static USED: AtomicU8 = AtomicU8::new(0);
+static USED_CS: crate::task::CriticalSection = crate::task::CriticalSection::new();
+
+/// SPI Host driver for SD cards supporting the SPI protocol.
+pub struct SdSpiHostDriver<'d, T> {
+    _spi_driver: T,
+    handle: sdspi_dev_handle_t,
+    _p: PhantomData<&'d mut ()>,
+}
+
+impl<'d, T> SdSpiHostDriver<'d, T>
+where
+    T: Borrow<SpiDriver<'d>>,
+{
+    /// Create a new SPI host driver for SD cards
+    ///
+    /// # Arguments
+    /// - spi_driver: SPI peripheral driver
+    /// - cs: Chip Select pin (optional)
+    /// - cd: Card Detect pin (optional)
+    /// - wp: Write Protect pin (optional)
+    /// - int: Interrupt pin (optional)
+    /// - wp_active_high: Write Protect active when high (optional, default = `false`)
+    pub fn new(
+        spi_driver: T,
+        cs: Option<impl Peripheral<P = impl OutputPin> + 'd>,
+        cd: Option<impl Peripheral<P = impl InputPin> + 'd>,
+        wp: Option<impl Peripheral<P = impl InputPin> + 'd>,
+        int: Option<impl Peripheral<P = impl InputPin> + 'd>,
+        #[cfg(not(any(
+            esp_idf_version_major = "4",
+            all(esp_idf_version_major = "5", esp_idf_version_minor = "0"),
+            all(esp_idf_version_major = "5", esp_idf_version_minor = "1"),
+        )))] // For ESP-IDF v5.2 and later
+        wp_active_high: Option<bool>,
```

Comment:
Option<bool> is a bit wired, can we make that an enum or something representing the polarity. I mean it's not really a Optional value ;D
 
 per docs: 
 "0 means "active low", i.e. card is protected when the GPIO is low; 1 means "active high", i.e. card is protected when GPIO is high"

Reply:
maybe we combine it with the wp pin itself ?

Reply:
It is an option because, well, it is not meaningful if the pin is None. Putting them behind a common Option is also an option, but then we get a nameless tuple which is not ideal, and naming the tuple with a custom struct is imo an overkill
 Anyway I'll see what to do.

Reply:
yeah directly combining is not nicely possible since its only included since 5.2..

Reply:
I think I'll keep it like that, even if a bit weird... :(